### PR TITLE
[dist, trainer, docs, ci] feat: support Qwen3-VL ChunkMBS with SP and EP

### PR DESCRIPTION
**Pull request by WenzheWang**
### What does this PR do?

> Extend the packed ChunkMBS support introduced in #898 to Qwen3-VL dense Ulysses SP and Qwen3-VL-MoE EP, including the combined SP+EP topology. This makes ChunkMBS's larger-outer-MBS and lower-gradient-accumulation strategy available to these distributed Qwen3-VL paths while preserving packed-sample boundaries, collective ordering, recomputation, and numerical behavior. The feature remains disabled by default.

### Checklist Before Starting

- Search for relative PRs/issues and link here: #898 is the merged ChunkMBS implementation baseline; no overlapping open Qwen3-VL SP/EP ChunkMBS PR was found.
- PR title follows `[{modules}] {type}: {description}` format.

### Test

- `.venv/bin/ruff check tasks tests veomni docs`: passed.
- `make quality`: passed (617 files checked and already formatted).
- `.venv/bin/python scripts/ci/check_doc_task_paths.py`: passed.
- `.venv/bin/pytest -q tests/parallel/test_chunk_mbs.py tests/distributed/test_gradient_checkpointing.py tests/utils/test_fsdp2_clip_grad_norm.py`: 92 passed, 7 skipped on macOS. The skipped cases are the existing real CPU FSDP2 cases plus the uneven dynamic-batch FSDP2/Ulysses case; Linux CI provides that coverage.
- Added Qwen3-VL dense SP and Qwen3-VL-MoE EP/SP+EP numerical checks for outputs, input gradients, parameter gradients, checkpoint recomputation, router logits, uneven dynamic batches, and FSDP chunk-count synchronization.
- Qwen3-VL E2E collection preserves the original non-ChunkMBS Dense/MoE cases and adds separate ChunkMBS-enabled cases.
- Independent `/veomni-review`: safe after the latest main refresh; hardware-only multi-device coverage remains for CI.

#### Numerical correctness and distributed coverage

NPU numerical training validation used feature commit `12b6a1d7` with Qwen3-VL-8B-Instruct, Ascend 910B2, PyTorch 2.10.0, torch-npu 2.10.0, and Transformers 5.9.0. For each lockstep step, the ChunkMBS pass and disabled reference pass used the same model state, input object, and RNG state; the input hash was checked after the ChunkMBS pass, and only the reference pass advanced AdamW. Tested head `9b2b54da` has the same runtime implementation and adds device-agnostic NPU test fixtures. Current head `afde6d1a` preserves the validated implementation and merges `origin/main` at `29e65e1c`.

| Scenario | Topology and controls | Result |
| --- | --- | --- |
| Dense dynamic batch with recomputation | 8 NPUs, DP-shard 4, Ulysses SP 2, MBS 8, GBS 64, ChunkMBS 2, non-reentrant checkpointing, 10 lockstep steps | max loss relative difference `2.60e-4`; max gradient-norm relative difference `1.37e-3`; gradient cosine >= `0.9992278`; AdamW-update cosine >= `0.9993028` |
| Dense fixed batch without recomputation | Same topology and batch sizes, 2 lockstep steps | loss matched exactly; max gradient relative L2 `9.99e-4`; gradient cosine >= `0.9999995`; AdamW-update cosine >= `0.9999783` |
| Qwen3-VL dense SP and MoE EP/SP+EP targeted regression | 4 NPUs; dense SP, MoE EP/SP+EP, checkpoint on/off, uneven dynamic ranges, and synchronized fallback | 13 passed in 300.20 seconds on `9b2b54da` |

An additional 12-condition Dense/MoE topology matrix covered SP1/SP2, EP1/EP2, dynamic batching, and non-reentrant recomputation. Every condition completed 10 optimizer steps, every enabled case performed real chunk work, and all predefined loss and gradient-norm gates passed.

#### Controlled performance evidence

The end-to-end strategy experiment used tested head `9b2b54da`, one node with 8 Ascend 910B2 NPUs, Qwen3-VL-8B-Instruct, BF16, FSDP2, PyTorch/torch-npu 2.10.0, Transformers 5.9.0, fixed input, `GBS=512`, seed 42, non-reentrant recomputation, a frozen vision encoder, and activation offload disabled. Each publishable condition used two independent repetitions, three warmup steps followed by nine measured steps, identical per-rank workload hashes, and zero allocator retries in the measured window.

| Comparison | Throughput | MFU | Interpretation |
| --- | ---: | ---: | --- |
| `MBS=8 / GA=8 / ChunkMBS off / RAF=True` -> `MBS=64 / GA=1 / ChunkMBS=16 / RAF=True` | **+28.52%** | **+9.98 pp** (`35.01%` -> `44.99%`) | ChunkMBS removes the per-layer working-set constraint that blocked the larger outer MBS and lower GA strategy |
| Default baseline -> `MBS=64 / GA=1 / ChunkMBS=16 / RAF=False` | **+30.05%** | **+10.52 pp** (`35.01%` -> `45.53%`) | Maximum measured end-to-end combination; disabling RAF adds only about 1.19% throughput and has a substantial memory cost |
| Dynamic-batch same-configuration ablation, ChunkMBS off -> 16 | `-2.00%` | `-0.91 pp` | With MBS, GA, inputs, topology, and RAF fixed, ChunkMBS reduced reserved memory by 2.86 GB with limited direct overhead |

The headline gains are practical batching-strategy comparisons, not a claim that ChunkMBS directly accelerates the same outer configuration. Their value comes from enabling a larger outer MBS and fewer gradient-accumulation rounds. Same-MBS controls also showed allocator headroom improvements: reserved memory decreased by about 3.65 GB for SP1 and 6.10 GB for SP2, while SP2 paid a larger redistribution overhead. Capacity-bound unchunked controls with allocator retries are retained only as capacity evidence and are excluded from the speedup claim.

### API and Usage Example

```yaml
train:
  accelerator:
    ulysses_size: 2
    ep_size: 2  # Qwen3-VL-MoE only
  chunk_mbs_config:
    enable: true
    chunk_mbs: 2
  gradient_checkpointing:
    enable_reentrant: false
```

`chunk_mbs` is the target number of packed samples per decoder-layer chunk. FSDP participants synchronize to the smallest non-empty chunk count for uneven dynamic batches, so a local chunk may contain more than this target. If any participant has only one chunk, the participating mesh uses the unchunked decoder path for that micro-batch.

Asynchronous Ulysses, context parallelism, TP, PP, other ExtraParallel/MoE models, `torch.compile`, `pad_to_length`, DPO, RL, and the custom Omni training loop remain unsupported.

### Design & Code Changes

> - Reuse Qwen3-VL decoder classes from `_no_split_modules`; no model-specific generated file is modified.
> - Reshard each packed chunk across Ulysses ranks, pad only the temporary rank-local shard, and restore outputs to the original sequence layout.
> - Restore Qwen3-VL-MoE router logits to their original layout so temporary Ulysses padding does not affect auxiliary load-balancing loss.
> - Synchronize chunk counts once per micro-batch across the owning FSDP or EP-FSDP mesh to preserve collective ordering for uneven dynamic batches; the mesh may already contain fused SP ranks.
> - Preserve non-reentrant gradient-checkpoint recomputation and support both dynamic and fixed packed batches.
> - Document the supported topology and explicit rejection boundaries.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- [x] Applied pre-commit checks
- [x] Added/updated documentation
- [x] Added/updated tests covered by the existing CI workflows

### Maintenance update (2026-08-23)

- Current head: `c402bb201d5d73d370f6541842fe9bfc8a7cc6c3`.
- Review follow-up preserves Qwen3-VL-MoE router outputs per sparse layer across chunks and follows the effective `output_router_logits` configuration for dict and tuple returns.
- Current-head validation: complete CPU ChunkMBS suite 71 passed and 7 skipped, including real tiny Qwen3-VL-MoE CausalLM aux-loss, mixed dense/sparse, flag-default, return-shape, and gradient coverage; `make quality` passed; no generated model file changed.
- All four review threads are resolved. Existing NPU/GPU performance evidence remains attached to its original head; no current-head GPU claim is made.
- Draft #1090 is a landing-order and config/docs/test conflict risk, not a replacement for the core `veomni/distributed/chunk_mbs.py` implementation. If #1090 lands first, this PR should rebase and reconcile only the overlapping support files.

## Summary by CodeRabbit

- **New Features**
  - Added ChunkMBS support for Qwen3-VL dense and MoE models.
  - Added compatibility with Ulysses sequence parallelism, expert parallelism, FSDP2, and uneven dynamic batches.
  - Added automatic chunk planning, padding, redistribution, and output restoration for supported configurations.
  - Preserved MoE router outputs and auxiliary losses across chunked training.

- **Bug Fixes**
  - Improved synchronization across distributed training meshes.
  - Added clearer validation for unsupported parallelism and checkpointing configurations.

- **Documentation**
  - Updated usage guidance, supported configurations, and known limitations.

### CODE REVIEW

**Comment by WenzheWang** [comment]
@Luosuu This follow-up extends #898 to Qwen3-VL dense Ulysses SP and Qwen3-VL-MoE EP, including SP+EP. Current head is `f96091fd`, with the latest `origin/main` (`1f586482`) merged cleanly.

I have added the controlled performance evidence requested for ChunkMBS. On 8 Ascend 910B2 NPUs with fixed `GBS=512`, moving from `MBS=8 / GA=8 / ChunkMBS off` to `MBS=64 / GA=1 / ChunkMBS=16` improved throughput by **28.52%** and MFU by **9.98 pp** with `reshard_after_forward=True`. The maximum measured combination with `reshard_after_forward=False` reached **30.05%** higher throughput and **10.52 pp** higher MFU. The PR body also reports the strict same-configuration ablation separately, so the larger-MBS/lower-GA benefit is not conflated with direct same-shape overhead.

The merged head passes Ruff and the refreshed local targeted suite (92 passed, 7 skipped). Could you please review the PR and approve the required workflows when convenient? Thanks.

**Comment by WenzheWang** [comment]
@Luosuu I refreshed the PR description with the completed controlled experiments and the evidence boundary.

The main benefit is now explicit: on 8 Ascend 910B2 NPUs with fixed `GBS=512`, ChunkMBS enabled `MBS=64 / GA=1` instead of the `MBS=8 / GA=8` baseline. With `reshard_after_forward=True`, throughput improved by **28.52%** and MFU increased by **9.98 pp**. The maximum measured combination with `reshard_after_forward=False` reached **30.05%** higher throughput and **10.52 pp** higher MFU, although RAF contributed only about 1.19% additional throughput and had a substantial memory cost.

The strict same-configuration controls are reported separately: the zero-retry dynamic-batch A-B-B-A comparison showed `-2.00%` throughput, `-0.91 pp` MFU, and `-2.86 GB` reserved memory with ChunkMBS enabled. This keeps the claim precise: ChunkMBS's advantage is enabling a larger outer MBS, fewer gradient-accumulation rounds, and more allocator headroom, rather than unconditional same-shape kernel acceleration.

The PR-specific SP/EP scope is covered by the real Qwen3-VL dense SP2 lockstep run and the Qwen3-VL dense/MoE SP/EP numerical matrix, including dynamic batching, non-reentrant recomputation, uneven ranges, router logits, and synchronized FSDP/EP fallback. The NPU results remain attributed to tested head `9b2b54da`. Current head `f96091fd` preserves both PR commits unchanged, merges the latest `origin/main` at `1f586482`, and passes the refreshed local targeted suite (92 passed, 7 skipped).

**Comment by WenzheWang** [comment]
@Luosuu The branch is synchronized with the current upstream `main` (`29e65e1c`; current head `afde6d1a`). The latest review feedback has been addressed: the per-micro-batch synchronization cost and mesh invariant are documented, Qwen3-VL compatibility-alias handling is explicitly scoped, and Dense/MoE E2E collection now preserves separate ChunkMBS-disabled and ChunkMBS-enabled cases. All review threads are resolved.

`make quality`, the documentation task-path check, and the targeted local suite pass (92 passed, 7 skipped). The PR description retains the controlled performance results with their tested-commit provenance: the larger-MBS/lower-GA comparison shows **+28.52% throughput and +9.98 pp MFU** with `reshard_after_forward=True`, while the maximum measured combination reaches **+30.05% throughput and +10.52 pp MFU**. The fork workflows are currently awaiting approval; could you please approve them and review the updated PR when convenient? Thanks.

File: veomni/distributed/chunk_mbs.py [hunk 1]
```diff
@@ -117,6 +145,50 @@ def build_chunk_mbs_ranges(batch: dict[str, Any], config: Any) -> Optional[list[
...
+        mesh = getattr(parallel_state, "fsdp_mesh", None)
+        if mesh is None:
+            raise RuntimeError("ChunkMBS with FSDP requires an initialized FSDP device mesh.")
+    if mesh is None:
+        return local_chunk_count
+
+    chunk_count = torch.tensor(local_chunk_count, dtype=torch.int32, device=get_device_type())
+    for mesh_dim_name in mesh.mesh_dim_names:
+        dist.all_reduce(chunk_count, op=dist.ReduceOp.MIN, group=mesh.get_group(mesh_dim_name))
+    return int(chunk_count.item())
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/distributed/chunk_mbs.py:182]: Addressed in afde6d1a. The reduction remains once per micro-batch because dynamic batches can change segment counts between micro-batches, so optimizer-step caching is not correct. The code and docs now state the exact collective/device-sync cost and clarify that no separate Ulysses-group reduction is needed; the owning FSDP or EP-FSDP mesh may already include fused SP ranks.

File: veomni/distributed/chunk_mbs.py [hunk 2]
```diff
@@ -314,6 +436,276 @@ def _chunked_forward(
...
+    padding_length = plan.chunk_shard_length * world_size - (seq_range.token_end - seq_range.token_start)
+    for key in ("cu_seq_lens_q", "cu_seq_lens_k"):
+        if key in chunk_kwargs and chunk_kwargs[key] is not None:
+            chunk_kwargs[key] = _slice_sp_cu_seq_lens(chunk_kwargs[key], seq_range, padding_length)
+
+    for key in ("max_length_q", "max_length_k"):
+        if key in chunk_kwargs:
+            chunk_kwargs[key] = max(seq_range.max_length, padding_length)
+
+    chunk_kwargs.pop("linear_attn_cu_seq_lens_q", None)
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply veomni/distributed/chunk_mbs.py:750]: Addressed in afde6d1a. I kept the compatibility-alias removal but made the allowlist dependency explicit: this path is restricted to Qwen3-VL decoder classes, which have no linear attention, while the collator still emits the alias. Raising on any non-None value was tested and rejected because it breaks valid Qwen3-VL Ulysses runs.

File: tests/e2e/test_e2e_parallel.py [hunk 3]
```diff
@@ -536,8 +548,10 @@ def test_qwen3vl_parallel_align(
     rtol: float,
     atol: float,
     max_sp_size: int | None,
-    dummy_qwen3vl_dataset,
+    request: pytest.FixtureRequest,
 ):
+    chunk_mbs_enabled = model_name in {"qwen3vl", "qwen3vlmoe"}
```
[the remark this answers is not in the document]
Comment (WenzheWang) [reply tests/e2e/test_e2e_parallel.py]: Addressed in afde6d1a. Qwen3-VL Dense and MoE now each collect separate ChunkMBS-disabled and ChunkMBS-enabled E2E cases, preserving the pre-existing parallel coverage while adding feature coverage.

**Comment by WenzheWang** [comment]
@Luosuu @Coach257 Current head c402bb20 addresses the latest router-logit contract review without editing generated Patchgen files. ChunkMBS now preserves one complete router tensor per sparse layer and the effective config default for aux loss. Validation: make quality and the complete CPU ChunkMBS suite pass (71 passed, 7 skipped), including real Qwen3-VL-MoE CausalLM labels/aux-loss/gradient coverage. All 4 review threads are resolved. No new GPU job was started for this follow-up. Could you please advise the landing order relative to #1090 and re-review when convenient?

**Comment by Coach257** [comment]
> @Luosuu @Coach257 Current head [c402bb2](https://github.com/ByteDance-Seed/VeOmni/commit/c402bb201d5d73d370f6541842fe9bfc8a7cc6c3) addresses the latest router-logit contract review without editing generated Patchgen files. ChunkMBS now preserves one complete router tensor per sparse layer and the effective config default for aux loss. Validation: make quality and the complete CPU ChunkMBS suite pass (71 passed, 7 skipped), including real Qwen3-VL-MoE CausalLM labels/aux-loss/gradient coverage. All 4 review threads are resolved. No new GPU job was started for this follow-up. Could you please advise the landing order relative to #1090 and re-review when convenient?

#1090 is just a reorg of args. There will be small conflicts. Maybe merge #933 after #1090 merged.
